### PR TITLE
Add UT for powervs_cluster file

### PR DESCRIPTION
**Pull request by Karthik-K-N**
**What this PR does / why we need it**:
This PR adds UT for powervs_cluster.go files

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes # Partially fixes https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/issues/1818

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by Karthik-K-N** [comment]
@dharaneeshvrd @Amulyam24 Please take a look and let me know your opinion, I tried to add UT for individual functions just to get more coverage and cover all the cases, Though few of the cases will be covered in UT for various Recocile functions.
Should this be continued this way or cover multiple function in same UT?

**Comment by dharaneeshvrd** [comment]
I think we can continue this way, if something broke it will help identify the exact func, if we dependant on the Reconcile test, may need some level of debugging. Also if we only test the main Reconcile func, there maybe chances for minor func to get missed.

**Comment by Karthik-K-N** [comment]
> I think we can continue this way, if something broke it will help identify the exact func, if we dependant on the Reconcile test, may need some level of debugging. Also if we only test the main Reconcile func, there maybe chances for minor func to get missed.

Sure thanks for the feedback, Will continue this way

**Review by Karthik-K-N — COMMENTED** [review]
This PR is ready for review. Please take a look.

File: cloud/scope/powervs_cluster_test.go [hunk 1]
```diff
@@ -153,3 +158,1457 @@ func TestNewPowerVSClusterScope(t *testing.T) {
...
+		g.Expect(err).NotTo(BeNil())
+	})
+	t.Run("When CreateResourceInstance returns error", func(t *testing.T) {
+		g := NewWithT(t)
+		setup(t)
+		t.Cleanup(teardown)
+
+		clusterScopeParams := getPowerVSClusterScopeParams()
+		clusterScopeParams.ResourceControllerFactory = func() (resourcecontroller.ResourceController, error) {
+			mockResourceController.EXPECT().CreateResourceInstance(gomock.Any()).Return(nil, nil, fmt.Errorf("intentional error"))
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1542]: might be good to have a better error message for readability and to keep in sync with previously added UTs

File: cloud/scope/powervs_cluster_test.go [hunk 2]
```diff
@@ -153,3 +158,1457 @@ func TestNewPowerVSClusterScope(t *testing.T) {
...
+		clusterScope, err := NewPowerVSClusterScope(clusterScopeParams)
+		g.Expect(err).To(BeNil())
+		clusterScope.IBMPowerVSCluster.Spec.ServiceInstanceID = serviceInstanceID
+
+		requeue, err := clusterScope.ReconcilePowerVSServiceInstance()
+		g.Expect(err).ToNot(BeNil())
+		g.Expect(requeue).To(BeFalse())
+	})
+
+	t.Run("When service instance id is set and instance in failed state", func(t *testing.T) {
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1189]: this test case has been repeated?

**Review by Amulyam24 — COMMENTED** [review]
Thanks @Karthik-K-N!
Minor comments , overall LGTM.

File: cloud/scope/powervs_cluster_test.go [hunk 3]
```diff
@@ -153,3 +158,1457 @@ func TestNewPowerVSClusterScope(t *testing.T) {
...
+			},
+			expectedID: ptr.To("dhcpserverid"),
+		},
+	}
+
+	for _, tc := range testCases {
+		g := NewWithT(t)
+		t.Run(tc.name, func(_ *testing.T) {
+			dhcpServerID := tc.clusterScope.GetDHCPServerID()
+			g.Expect(utils.DereferencePointer(dhcpServerID)).To(Equal(utils.DereferencePointer(tc.expectedID)))
```
[the remark this answers is not in the document]
Comment (Karthik-K-N) [reply cloud/scope/powervs_cluster_test.go:296]: Agreed, Will create subsequent PR to address the same. or if we can avoid Dereference as it wont support all the data types as well.

File: cloud/scope/powervs_cluster_test.go [hunk 4]
```diff
@@ -153,3 +158,1457 @@ func TestNewPowerVSClusterScope(t *testing.T) {
...
+						ServiceInstance: &infrav1beta2.ResourceReference{
+							ID: ptr.To("statusServiceInstanceID"),
+						},
+					},
+				},
+			},
+			expectedID: "statusServiceInstanceID",
+		},
+		{
+			name: "Spec Service Instance ID takes precedence over status Service Instance ID",
```
[the remark this answers is not in the document]
Comment (Karthik-K-N) [reply cloud/scope/powervs_cluster_test.go:212]: As of today it wont, added it so the future refactor of code should not break the flow. We need to consider one case where if user sets spec.id and what if there is no resource exist with that id the controller goes and creates a resource with default name format.

File: cloud/scope/powervs_cluster_test.go (hunk 2 above)
Comment (Karthik-K-N) [reply cloud/scope/powervs_cluster_test.go]: Yes, Removed duplicate

File: cloud/scope/powervs_cluster_test.go [hunk 5]
```diff
@@ -153,3 +158,1457 @@ func TestNewPowerVSClusterScope(t *testing.T) {
...
+			lbName:     "lb",
+			expectedID: ptr.To("lb-1"),
+		},
+	}
+
+	for _, tc := range testCases {
+		g := NewWithT(t)
+		t.Run(tc.name, func(_ *testing.T) {
+			lbID := tc.clusterScope.GetLoadBalancerID(tc.lbName)
+			fmt.Println(lbID)
```
[the remark this answers is not in the document]
Comment (Karthik-K-N) [reply cloud/scope/powervs_cluster_test.go]: Removed, Added while debugging.

File: cloud/scope/powervs_cluster_test.go [hunk 6]
```diff
@@ -153,3 +159,1432 @@ func TestNewPowerVSClusterScope(t *testing.T) {
...
+
+		clusterScopeParams := getPowerVSClusterScopeParams()
+		clusterScopeParams.ResourceControllerFactory = func() (resourcecontroller.ResourceController, error) {
+			mockResourceController.EXPECT().GetServiceInstance(gomock.Any(), gomock.Any(), gomock.Any()).Return(nil, errors.New("failed to get service instance"))
+			return mockResourceController, nil
+		}
+		clusterScope, err := NewPowerVSClusterScope(clusterScopeParams)
+		g.Expect(err).To(BeNil())
+
+		instance, requeue, err := clusterScope.isServiceInstanceExists()
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1391]:
```suggestion
		instanceID, requeue, err := clusterScope.isServiceInstanceExists()
```
please change this wherever applicable in this test case

Comment (Karthik-K-N) [reply cloud/scope/powervs_cluster_test.go]: Updated at all the places.

File: cloud/scope/powervs_cluster_test.go [hunk 7]
```diff
@@ -153,3 +159,1432 @@ func TestNewPowerVSClusterScope(t *testing.T) {
...
+			lbName:        "lb",
+			expectedState: ptr.To(infrav1beta2.VPCLoadBalancerStateActive),
+		},
+	}
+
+	for _, tc := range testCases {
+		g := NewWithT(t)
+		t.Run(tc.name, func(_ *testing.T) {
+			lbState := tc.clusterScope.GetLoadBalancerState(tc.lbName)
+			fmt.Println(lbState)
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:877]: If not needed, can be removed.

File: cloud/scope/powervs_cluster_test.go [hunk 8]
```diff
@@ -153,3 +159,1432 @@ func TestNewPowerVSClusterScope(t *testing.T) {
...
+		clusterScopeParams := getPowerVSClusterScopeParams()
+		clusterScopeParams.ResourceControllerFactory = func() (resourcecontroller.ResourceController, error) {
+			mockResourceController.EXPECT().CreateResourceInstance(gomock.Any()).Return(nil, nil, errors.New("failed to create resource instance"))
+			return mockResourceController, nil
+		}
+		clusterScope, err := NewPowerVSClusterScope(clusterScopeParams)
+		g.Expect(err).To(BeNil())
+
+		instance, err := clusterScope.createServiceInstance()
+		fmt.Println(err)
```
Comment (Amulyam24) [inline cloud/scope/powervs_cluster_test.go:1525]: If not needed, can be removed.

**Review by Amulyam24 — COMMENTED** [review]
LGTM! 
minor nits

**Review by Amulyam24 — COMMENTED** [review]
/lgtm

**Comment by Karthik-K-N** [comment]
Updated and rebased with main, PR is ready for review again.

**Review by Amulyam24 — COMMENTED** [review]
/lgtm

**Review by mkumatag — APPROVED** [review]

**Pull request merged**